formver.compiler-plugin/core/src/org/jetbrains/kotlin/formver/core/linearization/SsaConverter.kt

```diff
@@ -2,33 +2,49 @@ package org.jetbrains.kotlin.formver.core.linearization
 
 import org.jetbrains.kotlin.KtSourceElement
 import org.jetbrains.kotlin.formver.common.SnaktInternalException
+import org.jetbrains.kotlin.formver.core.names.SsaVariableName
+import org.jetbrains.kotlin.formver.viper.SymbolicName
 import org.jetbrains.kotlin.formver.viper.ast.Declaration
 import org.jetbrains.kotlin.formver.viper.ast.Exp
 
-class SsaConverter(val source: KtSourceElement? = null) {
-    private val assignments: MutableMap<Declaration.LocalVarDecl, Exp> = mutableMapOf()
+class SsaConverter(
+    val source: KtSourceElement? = null,
+    val variableIndex: MutableMap<SymbolicName, Int> = mutableMapOf()
+) {
+    private val assignments: MutableList<SsaAssignment> = mutableListOf()
     private var body: Exp? = null
 
     fun asExp(): Exp {
         if (body == null) throw SnaktInternalException(
             source,
             "Empty body cannot be converted into let chain"
         )
-        return assignments.entries.toList().foldRight(body as Exp) { (decl, expr), acc ->
-            Exp.LetBinding(decl, expr, acc)
+        val bodyExp = body!!
+        return assignments.foldRight(bodyExp) { (decl, ssaIdx, expr), acc ->
+            Exp.LetBinding(decl.copy(name = SsaVariableName(decl.name, ssaIdx)), expr, acc)
         }
     }
 
     fun addAssignment(decl: Declaration.LocalVarDecl, varExp: Exp) {
-        // TODO: Allow this case using SSA transformation
-        if (assignments.containsKey(decl)) throw SnaktInternalException(
-            source,
-            "Found duplicate variable declaration while constructing a let-chain. Declaration is $decl"
-        )
-        assignments[decl] = varExp
+        variableIndex[decl.name] = variableIndex[decl.name]?.plus(1) ?: 0
+        assignments.add(SsaAssignment(decl, variableIndex[decl.name] as Int, varExp))
     }
 
     fun addBody(body: Exp) {
         this.body = body
     }
-}
\ No newline at end of file
+
+    /**
+     * Resolves a base variable name to the latest SSA variable name used in this or previous converters.
+     * The name is resolved as follows:
```

Comment:
In general, the point of a comment is to explain why things are the way they are; what happens should be clear from the code.